lib/SecretsPluginApi/extensionplugins.cpp

```diff
@@ -13,71 +13,1189 @@
 
 using namespace Sailfish::Secrets;
 
+/*!
+  \class PluginBase
+  \brief Provides the base interface for extension plugins for the Sailfish OS Secrets and Crypto Framework.
+
+  PluginBase is an abstract base class which provides an interface which all extension plugins
+  must implement.  This interface includes some information about the plugin (including its
+  name and version, and whether or not it supports locking), as well as locking operation methods.
+
+  Extension plugins are loaded by the Sailfish OS Secrets and Crypto Framework daemon, and
+  the functionality of those plugins is exposed to clients via the Secrets and Crypto API.
+
+  There are a variety of extension plugin types which each extend PluginBase,
+  which define specific interfaces which concrete derived types must implement.
+  A list of the available Secrets extension plugin types follows:
+
+  \list
+  \li \l{Sailfish::Secrets::AuthenticationPlugin} perform suser authentication and verification
+  \li \l{Sailfish::Secrets::EncryptionPlugin} encrypts secret data for secrets stored in a \l{Sailfish::Secrets::StoragePlugin}
+  \li \l{Sailfish::Secrets::StoragePlugin} provides unencrypted storage for secrets
+  \li \l{Sailfish::Secrets::EncryptedStoragePlugin} provides encrypted storage for secrets
+  \endlist
+
+  In addition, there also exists the \l{Sailfish::Crypto::CryptoPlugin} extension plugin type,
+  which provides cryptographic functionality to clients.  If a particular plugin implements
+  both the \l{Sailfish::Crypto::CryptoPlugin} and \l{Sailfish::Secrets::EncryptedStoragePlugin}
+  interfaces then it is able to secure store cryptographic keys as well as perform cryptographic
+  operations using those keys, and is referred to as a Crypto-Storage Plugin.
+
+  Plugin implementers should not extend PluginBase directly, but should instead derive from
+  one of the above-listed plugin types (or both CryptoPlugin and EncryptedStoragePlugin if
+  they are implementing a Crypto-Storage plugin).
+
+  Plugin implementers must be aware that the information reporting methods (name(), version(),
+  and supportsLocking()) will be invoked from the main thread of the secrets daemon, while
+  the various locking operation methods (isLocked(), lock(), unlock(), and setLockCode())
+  will be invoked from a separate thread.  Plugins are loaded and plugin instances are constructed
+  in the main thread.
+
+  In order to implement a Secrets extension plugin, plugin implementers should
+  specify the following in their .pro file:
+  \code
+  CONFIG += link_pkgconfig
+  PKGCONFIG += sailfishsecretspluginapi
+  \endcode
+ */
+
+/*!
+ * \fn PluginBase::name() const
+ * \brief Return the name of the plugin
+ *
+ * This name must be globally unique, so a fully-qualified-domain-name
+ * prefix is recommended.
```

Comment:
Give an example here, it will be clearer.

Reply:
Done